hibernate-core/src/main/java/org/hibernate/sql/ast/spi/StandardSqlAstInsertTranslator.java

```diff
@@ -42,37 +43,62 @@ public JdbcInsert translate(InsertSelectStatement sqlAst) {
 			renderImplicitTargetColumnSpec();
 		}
 		else {
-			for ( int i = 0; i < targetColumnReferences.size(); i++ ) {
-				if ( firstPass ) {
+			for (ColumnReference targetColumnReference : targetColumnReferences) {
+				if (firstPass) {
 					firstPass = false;
 				}
 				else {
 					appendSql( ", " );
 				}
 
-				final ColumnReference columnReference = targetColumnReferences.get( i );
-				appendSql( columnReference.getColumnExpression() );
+				appendSql( targetColumnReference.getColumnExpression() );
 			}
 		}
 
 		appendSql( ") " );
 
-		visitQuerySpec( sqlAst.getSourceSelectStatement() );
+		if ( sqlAst.getSourceSelectStatement()!=null ) {
+			visitQuerySpec( sqlAst.getSourceSelectStatement() );
+		}
+		else {
+			appendSql("values");
+			boolean firstTuple = true;
```

Comment:
Do we need to check the values and the columns share the same length here? Seems better to initiate a SQL roundtrip which is destined to fail.

Reply:
Definitely don't want to do it here in the code that renders SQL, no.